src/main/java/nextstep/courses/domain/SessionStudent.java

```diff
@@ -2,34 +2,55 @@
 
 import nextstep.users.domain.NsUser;
 
-import java.util.ArrayList;
-import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
 
 public class SessionStudent {
 
     private final int availableSeats;
-    private List<NsUser> students = new ArrayList<NsUser>();
+    private final Set<Long> studentIds;
 
-    public SessionStudent(int availableSeats) {
+    public SessionStudent(int availableSeats){
+        this(availableSeats, new HashSet<>());
+    }
+
+    public SessionStudent(int availableSeats, Set<Long> studentIds) {
         this.availableSeats = availableSeats;
+        this.studentIds = studentIds;
     }
 
     public void addStudent(Pricing pricing, NsUser user) {
         checkAddStudent(pricing, user);
-        students.add(user);
+        studentIds.add(user.getId());
     }
 
-    public int size(){
-        return students.size();
+    public int size() {
+        return studentIds.size();
     }
 
     private void checkAddStudent(Pricing pricing, NsUser user) {
         if (pricing.isFree()) {
             return;
         }
+        checkFullSession();
+        checkExistingStudent(user);
+    }
 
-        if (students.size() == availableSeats) {
+    private void checkFullSession() {
+        if (studentIds.size() == availableSeats) {
             throw new IllegalStateException("이 강의는 정원이 초과되었습니다.");
         }
     }
+
+    private void checkExistingStudent(NsUser user) {
+        boolean isExistingStudent = studentIds.stream()
+                .anyMatch(student -> student.equals(user.getId()));
+        if (isExistingStudent) {
```

Comment:
Set의 `contains()`를 활용해보셔도 좋을 것 같습니다.
```suggestion
        if (studentIds.contains(user.getId())) {
```